lib/db.js

```diff
@@ -0,0 +1,181 @@
+    }
+
+    return obj;
+  });
+}
+
+function streamRecordsToPg (stream, pg) {
+  const table = pg('measurements')
+    .returning('location');
+
+  return stream
+    .tap()
+    .pipe(new DataStream())
+    .setOptions({maxParallel: 1})
+    .use(convertMeasurementToSQLObject, pg)
+    .assign(async record => {
+      try {
+        await table.insert(record);
+      } catch (cause) {
+        if (cause.code === '23505') {
+          return { status: 'duplicate' };
+        }
+
+        throw cause;
+      }
+      return { status: 'inserted' };
+    });
+}
+
+async function streamDataToDB (stream, pg, counts) {
+  return stream
+    .use(streamRecordsToPg, pg)
+    .filter(({status}) => {
+      if (status === 'duplicate') {
+        counts.duplicates++;
+        return false;
+      }
+
+      counts.inserted++;
+      return true;
+    })
+    .reduce(x => x++, 0)
+  ;
+}
+
+async function saveResultsToS3 (stream, s3, bucketName, key) {
+  return stream
+    .flatMap(({results}) => results)
+    .map(result => result.data)
+    .use(streamDataToS3, s3, bucketName, key)
+  ;
+}
+
+async function streamDataToS3 (stream, s3, bucketName, key) {
+  return promisify(s3.upload, {
+    Body: stream.JSONStringify('\n'),
+    Bucket: bucketName,
+    Key: key
+  });
+}
+
+/**
+ * Saves information about fetches to the database
+ *
+ * @param {FetchResult} sources
+ */
+export async function saveFetches ({timeStarted, timeEnded, itemsInserted, errors, results}) {
+  const pg = await getDB();
+
+  return pg('fetches')
+    .insert([{
+      time_started: new Date(timeStarted),
+      time_ended: new Date(timeEnded),
+      count: itemsInserted,
+      results: JSON.stringify(errors || results)
+    }])
+    .then(() => log.info('Fetches table successfully updated'))
+    .catch((e) => log.error(e));
+}
+
+/**
+ * Save sources information, overwritten any previous results
+ *
+ * @param {Sources} sources
+ */
+export async function saveSources (sources) {
+  const inserts = Object.values(sources)
+    .map(data => ({data: JSON.stringify(data)}));
+
+  const pg = await getDB();
+  return pg('sources')
+    .del()
+    .then(() => log.verbose('Sources table successfully deleted.'))
+    .then(() => pg('sources').insert(inserts))
+    .then(() => log.info('Sources table successfully updated.'))
+    .catch((e) => log.error(e));
+}
+
+/**
+ * Streams measurements to database and cloud storage
+ *
+ * @param {OpenAQEnv} env
+ * @param {String} bucketName
+ */
+export function streamMeasurementsToDBAndStorage ({doSaveToS3, dryrun}, bucketName) {
+  return async ({ stream, counts }) => {
+    if (dryrun) {
+      return stream
+        .do(m => log.verbose(JSON.stringify(m)))
+        .run();
+    } else {
+      const key = `realtime/${moment().format('YYYY-MM-DD/X')}.ndjson`;
+      const pg = await getDB();
+
+      await Promise.all([
+        doSaveToS3 ? saveResultsToS3(stream, new (require('aws-sdk').S3)(), bucketName, key) : true,
```

Comment:
Was looking to test the data format of data saved to S3. Ran with the following command (using valid AWS key/secret)

`AWS_ACCESS_KEY_ID=ID AWS_SECRET_ACCESS_KEY=KEY AWS_BUCKET_NAME=openaq-fetches SAVE_TO_S3=true node index.js` 

and was seeing a lot of the following errors

```
(node:29353) UnhandledPromiseRejectionWarning: Error: Cannot read property 'Symbol(Symbol.iterator)' of undefined
    at promise.catch (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/util/promise-transform-stream.js:221:31)
    at process._tickCallback (internal/process/next_tick.js:178:7)
  caused by:
    at into (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet/lib/data-stream.js:380:35)
    at process._tickCallback (internal/process/next_tick.js:178:7)
  --- raised in DataStream(unused) constructed ---
    at new PromiseTransformStream (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/util/promise-transform-stream.js:65:27)
    at new DataStream (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/data-stream.js:55:9)
    at DataStream.flatMap (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet/lib/data-stream.js:373:21)
    at saveResultsToS3 (/Users/jflasher/workspace/openaq-fetch/lib/db.js:108:6)
    at /Users/jflasher/workspace/openaq-fetch/lib/db.js:176:22
    at process._tickCallback (internal/process/next_tick.js:178:7)
(node:29353) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 186533)
```

Tried again without the extra `SAVE_TO_DB` and `AWS_BUCKET_NAME` vars and everything worked fine again, so I presume it's tied to the output there. What's interesting is that this didn't happen at the end of the fetch which is when I presume the data is getting written to S3, but happened in mid-fetch as the adapters were running.

Reply:
The happening in middle part is expected - steaming to s3 happens white adapters were still running. I'll check this out tomorrow.

Reply:
Ok. For some reason I didn't merge some of the work from the s3 branch and that was an old version which didn't have a slight chance to work.

@jflasher - there's a limitation of s3 that requires the uploads to have file sizes set before the upload - trouble is, we don't actually know how big the file will be. Therefore the stream is chunked. In the upload code I set a chunk to 1MB - which sets a maximum of 10GB per file (S3 allows a maximum of 10k chunks).

Now I thought this should be controller via env, so let me add a quick commit for that.

Reply:
Ok, that's also pushed.

Reply:
Yeah. I did note that. So I'm rewriting the save to s3 so that all the measurements are sent to a single stream constructed before the fetch is started - and that stream will get saved.

I got this working, but at the same time I'll need to spend little more time to make it look better.